Update Performance page with notes from the team (#1)
amcasey/TypeScript_wiki: Performance.md

+ }
```

## Using Type Annotations

Adding type annotations, especially return types, can save the compiler a lot of work.
In part, this is because named types tend to be more compact than anonymous types (which the compiler might infer), which reduces the amount of time spend reading and writing declaration files (e.g. for incremental builds).
Type inference is very convenient, so there's no need to do this universally - however, it can be a useful thing to try if you've identified a slow section of your code.

```diff

> - import { otherFunc } from "other";
> -
> - export function func() {
> -     return otherFunc();
> - }
> + import { otherFunc, otherType } from "other";
> +
> + export function func(): otherType {
> +     return otherFunc();
> + }

Reviewer: ```suggestion
- import { otherFunc } from "other";
+ import { otherFunc, otherType } from "other";
 
- export function func() {
+ export function func(): otherType {
      return otherFunc();
  }
```

which will look like

```diff
- import { otherFunc } from "other";
+ import { otherFunc, otherType } from "other";
 
- export function func() {
+ export function func(): otherType {
      return otherFunc();
  }
```

```

## Preferring Base Types Over Unions

Union types are great - they let you express the range of possible values for a type.

```ts
interface WeekdaySchedule {
  day: "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday";
  wake: Time;
  startWork: Time;
  endWork: Time;
  sleep: Time;
}

interface WeekendSchedule {
  day: "Saturday" | "Sunday";
  wake: Time;
  familyMeal: Time;
  sleep: Time;
}

declare function printSchedule(schedule: WeekdaySchedule | WeekendSchedule);
```

However, they also come with a cost.

> Every time an object is assigned to a `Schedule`, it has to be compared to each element of the union.

Reviewer: there's no `Schedule` defined

Author: Oops, edited my example.

For a two-element union, this is trivial and inexpensive.

> However, if your union has more than a dozen elements, it can cause real problems.

Reviewer: ```suggestion
However, if your union has more than a dozen elements, it can cause real problems in compilation speed.
```

For instance, to eliminate redundant members from a union, the elements have to be compared pairwise, which is quadratic.

> Similarly, if you intersect large unions, distribution can result in enormous types.

Reviewer: ```suggestion
This sort of check might occur when intersecting large unions, where intersecting over each union member can result in enormous types that then need to be reduced.
```

One way to avoid this is to use subtypes, rather than unions.

```ts
interface Schedule {
  day: "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday";
  wake: Time;
  sleep: Time;
}

interface WeekdaySchedule extends Schedule {
  day: "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday";
  startWork: Time;
  endWork: Time;
}

interface WeekendSchedule extends Schedule {
  day: "Saturday" | "Sunday";
  familyMeal: Time;
}

declare function printSchedule(schedule: Schedule);
```

> More realistically, consider declaring a base `HtmlElement` type, which `DivElement`, `ImgElement`, etc extend, rather than using a union `DivElement | /*...*/ | ImgElement | /*...*/`.

Reviewer: ```suggestion
A more realistic example of this might come up when trying to model every built-in DOM element type.
In this case, it would be preferable to create a base `HtmlElement` type with common members, which `DivElement`, `ImgElement`, etc. extend, rather than to create the exhaustive union `DivElement | /*...*/ | ImgElement | /*...*/`.
```

------------                ------------
```

One commonly asked question is "how big should a project be?".

> This is a lot like asking "how big should a function be?" or "how big should a class be?" and, to a large extent, comes down to experience.

Reviewer: ```suggestion
This is a lot like asking "how big should a function be?" or "how big should a class be?" and, to a large extent, it comes down to experience.
```

These two options can often hide misconfiguration and conflicts in `.d.ts` files, so we suggest using them *only* for faster builds.

## Using Faster Variance Checks

Is a list of dogs a list of animals?
That is, is `List<Dog>` assignable to `List<Animals>`?
The straightforward way to find out is to do a structural comparison of the types, member by member.
Unfortunately, this can be very expensive.
However, if we know enough about `List<T>`, we can reduce this assignability check to determining whether `Dog` is assignable to `Animal` (i.e. without considering each member of `List<T>`).
(In particular, we need to know the [variance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) of the type parameter `T`.)

> The compiler can only take full advantage of this potential speedup if the `strictFunctionTypes` flag is enabled (otherwise, it uses the slower, but more lenient, structural check).

Reviewer: It's unclear what the action item is for this one at a glance. Maybe add

```md
For this reason, we recommend building with `--strictFunctionTypes` (which is enabled by default under `--strict`).
```

Author: Not sure why this one wasn't a suggested edit. 😛
